### PR TITLE
Introduce "dynamic_template" property for fields

### LINKED ISSUE #187: Add dynamic_template to field properties
See https://github.com/elastic/package-spec/issues/171. Add `dynamic_template` to the field schema, which will enable packages to define named dynamic templates.

### PR DESCRIPTION
**Pull request by axw**
## What does this PR do?

This PR introduces a new "dynamic_template" property for fields.

## Why is it important?

See https://github.com/elastic/package-spec/issues/171

## Checklist

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/master/test/packages) that prove my change is effective.
- [x] I have added an entry in [`versions/N/changelog.yml`](https://github.com/elastic/package-spec/blob/master/versions/1/changelog.yml).

## Related issues

Closes #187

### CODE REVIEW

**Comment by axw** [comment]
CI says:

> Specs are not up to date. Please run make update.

... but when I run `make update`, it says the specs are up to date. Also, when I run `make check` I don't get any errors.

File: test/packages/bad_dynamic_template/changelog.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- version: 0.1.2
+  changes:
+    - description: initial release
+      type: enhancement
+      link: https://github.com/elastic/package-spec/pull/131
```
Comment (mtojek) [inline test/packages/bad_dynamic_template/changelog.yml:5]: we used to link it to spec's PR. In this case: https://github.com/elastic/package-spec/pull/188

File: versions/1/data_stream/fields/fields.spec.yml [hunk 2]
```diff
@@ -86,5 +86,8 @@ spec:
       fields:
         description: Sub-fields, when type is group
         $ref: "#"     # JSON-schema syntax for pointing to the root of the schema
+      dynamic_template:
```
Comment (mtojek) [inline versions/1/data_stream/fields/fields.spec.yml:89]: nit: could you please put it before `fields`? Let's leave the subgroup of fields as the last item here.

**Review by mtojek — COMMENTED** [review]
> ... but when I run make update, it says the specs are up to date. Also, when I run make check I don't get any errors.

Not sure what went wrong in your case, but I've observed the same behavior as CI on my end (and pushed a fix to this PR).

File: test/packages/bad_dynamic_template/changelog.yml (hunk 1 above)
Comment (axw) [reply test/packages/bad_dynamic_template/changelog.yml]: Done.

**Comment by axw** [comment]
@mtojek FWIW I've tried running with Go 1.15.8, which is what CI appears to be using - no idea why `make update` is doing something different here...

**Comment by mtojek** [comment]
Is there an option that your IDE reformats the code?

**Comment by axw** [comment]
I'm not running in an IDE, I'm running `make update` from the terminal.

**Review by mtojek — APPROVED** [review]

**Comment by mtojek** [comment]
I compared both blobs (valid and invalid) and here are results.

valid:
```
0000b3a0: 305c 7830 305c 7830 305c 7830 305c 7862  0\x00\x00\x00\xb
0000b3b0: 345c 7838 315c 7830 305c 7830 305c 7830  4\x81\x00\x00\x0
0000b3c0: 305c 7830 3031 2f5f 6465 762f 6275 696c  0\x001/_dev/buil
0000b3d0: 642f 6275 696c 642e 7370 6563 2e79 6d6c  d/build.spec.yml
0000b3e0: 5554 5c78 3035 5c78 3030 5c78 3031 5c78  UT\x05\x00\x01\x
0000b3f0: 3830 436d 3850 4b5c 7830 315c 7830 325c  80Cm8PK\x01\x02\
0000b400: 7831 345c 7830 335c 7831 345c 7830 305c  x14\x03\x14\x00\
0000b410: 7830 385c 7830 305c 7830 385c 7830 305c  x08\x00\x08\x00\
0000b420: 7830 305c 7830 3021 285c 7831 653b 5c78  x00\x00!(\x1e;\x
```

invalid:
```
0000b3a0: 305c 7830 305c 7830 305c 7830 305c 7861  0\x00\x00\x00\xa
0000b3b0: 345c 7838 315c 7830 305c 7830 305c 7830  4\x81\x00\x00\x0
0000b3c0: 305c 7830 3031 2f5f 6465 762f 6275 696c  0\x001/_dev/buil
0000b3d0: 642f 6275 696c 642e 7370 6563 2e79 6d6c  d/build.spec.yml
0000b3e0: 5554 5c78 3035 5c78 3030 5c78 3031 5c78  UT\x05\x00\x01\x
0000b3f0: 3830 436d 3850 4b5c 7830 315c 7830 325c  80Cm8PK\x01\x02\
0000b400: 7831 345c 7830 335c 7831 345c 7830 305c  x14\x03\x14\x00\
0000b410: 7830 385c 7830 305c 7830 385c 7830 305c  x08\x00\x08\x00\
0000b420: 7830 305c 7830 3021 285c 7831 653b 5c78  x00\x00!(\x1e;\x
```

there are small differences across the file: `B4 81 00` vs `A4 81 00`. I will try to dig deeper.

**Comment by mtojek** [comment]
It seems that difference comes from file attributes. Could you please share your chmods for:

```
versions/1/_dev/build/
versions/1/
```

In my case these are:

```
drwxr-xr-x  5 marcin.tojek  staff  160 16 cze 12:16 .
drwxr-xr-x  5 marcin.tojek  staff  160 18 lis  2020 ..
-rw-r--r--  1 marcin.tojek  staff  720 16 cze 12:16 build.spec.yml
drwxr-xr-x  3 marcin.tojek  staff   96  2 mar 13:40 docs
-rw-r--r--  1 marcin.tojek  staff  356 16 cze 12:16 spec.yml
```

```
drwxr-xr-x  11 marcin.tojek  staff   352 16 cze 12:16 .
drwxr-xr-x   3 marcin.tojek  staff    96 25 sie  2020 ..
drwxr-xr-x   5 marcin.tojek  staff   160 18 lis  2020 _dev
drwxr-xr-x   3 marcin.tojek  staff    96 19 lis  2020 agent
-rw-r--r--   1 marcin.tojek  staff  1262 27 kwi 12:23 changelog.spec.yml
-rw-r--r--   1 marcin.tojek  staff  6564 16 cze 12:16 changelog.yml
drwxr-xr-x   6 marcin.tojek  staff   192  2 lut 13:16 data_stream
drwxr-xr-x   3 marcin.tojek  staff    96  2 mar 13:40 docs
drwxr-xr-x   3 marcin.tojek  staff    96 27 kwi 12:23 kibana
-rw-r--r--   1 marcin.tojek  staff  8080 10 cze 15:03 manifest.spec.yml
-rw-r--r--   1 marcin.tojek  staff  1429 10 cze 10:16 spec.yml
```

BTW which operating system are you using?

**Comment by axw** [comment]
Ah. I'm running Ubuntu 20.04. I suppose we have different `umask` settings -- mine is 0002. Maybe we can update to Go 1.16.x and use go:embed instead of statik?

**Comment by mtojek** [comment]
Yes, we've already planned it: https://github.com/elastic/package-spec/issues/99

**Pull request merged**